### PR TITLE
Add type hint checking

### PR DESCRIPTION
**Pull request by bjgill**
Use mypy to check the types in this library. Then mark the library as having type hints so consumers can use them.

Note that the type hinting is thus far rather sparse, so this will not yet be very useful to consumers.

Copied very heavily from https://github.com/alphagov/digitalmarketplace-utils/pull/606

### CODE REVIEW

File: tasks.py [hunk 1]
```diff
@@ -1 +1,12 @@
+from invoke import task
+
 from dmdevtools.invoke_tasks import library_tasks as ns
+
+
+@task(ns["virtualenv"], ns["requirements_dev"])
+def test_mypy(c):
+    c.run("mypy dmutils/")
```
Comment (alex9smith) [inline tasks.py:8]:
```suggestion
    c.run("mypy dmapiclient/")
```
I _think_ this is the only reason why the checks are failing - mypy seems happy on my machine

Comment (lfdebrux) [reply tasks.py:8]: Might be worth instead adding a `mypy.ini` file

```
[mypy]
files = dmapiclient
```

Then running `mypy` should just do the right thing.
```

Comment (bjgill) [reply tasks.py:8]: Huh. I didn't know about that. We already have a `mypy.ini`, so that sounds most useful

Comment (bjgill) [reply tasks.py]: Fixed and rebased against main

**Comment by bjgill** [comment]
Oh dear - tests broken for an unrelated reason that's going to hit an number of other repos. I've created a ticket: https://trello.com/c/vRE4IeaF/2206-fix-github-actions-ci-for-repos-that-dont-have-pip-tools

**Review by lfdebrux — APPROVED** [review]

**Pull request merged**